crates/sdk/src/env/ext.rs

```diff
@@ -1,8 +1,120 @@
 use crate::sys;
-use crate::sys::Buffer;
+use crate::sys::{Buffer, BufferMut};
+
+/// A blockchain proposal action.
+///
+/// This enum represents the different actions that can be executed against a
+/// blockchain, and combined into a proposal.
+///
+#[derive(BorshDeserialize, BorshSerialize, Clone, Debug, Eq, PartialEq)]
+pub enum ProposalAction {
+    /// Call a method on a contract.
+    ExternalFunctionCall {
+        /// The account ID of the contract to call.
+        receiver_id: AccountId,
+
+        /// The method name to call.
+        method_name: String,
+
+        /// The arguments to pass to the method.
+        args: String,
+
+        /// The amount of tokens to attach to the call.
+        deposit: u64,
+
+        /// The maximum amount of gas to use for the call.
+        gas: u64,
+    },
+
+    /// Transfer tokens to an account.
+    Transfer {
+        /// The account ID of the receiver.
+        receiver_id: AccountId,
+
+        /// The amount of tokens to transfer.
+        amount: u64,
+    },
+
+    /// Set the number of approvals required for a proposal to be executed.
+    SetNumApprovals {
+        /// The number of approvals required.
+        num_approvals: u32,
+    },
+
+    /// Set the number of active proposals allowed at once.
+    SetActiveProposalsLimit {
+        /// The number of active proposals allowed.
+        active_proposals_limit: u32,
+    },
+
+    /// Set a value in the contract's context.
+    SetContextValue {
+        /// The key to set.
+        key: Box<[u8]>,
+
+        /// The value to set.
+        value: Box<[u8]>,
+    },
+}
+
+/// Unique identifier for an account.
+#[derive(BorshDeserialize, BorshSerialize, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct AccountId(String);
+
+/// A draft proposal.
+///
+/// This struct is used to build a proposal before sending it to the blockchain.
+/// It is distinct from a proposal that has been prepared and needs signing.
+///
+#[derive(BorshDeserialize, BorshSerialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct DraftProposal {
+    /// The actions to be executed by the proposal. One proposal can contain
+    /// multiple actions to execute.
+    actions: Vec<ProposalAction>,
+}
+
+impl DraftProposal {
+    /// Add an action to transfer tokens to an account.
+    #[must_use]
+    pub fn transfer(mut self, receiver: AccountId, amount: u64) -> Self {
+        self.actions.push(ProposalAction::Transfer {
+            receiver_id: receiver,
+            amount,
+        });
+        self
+    }
+
+    /// Finalise the proposal and send it to the blockchain.
+    #[must_use]
+    pub fn send(self) -> ProposalId {
+        #[expect(unused_mut, reason = "Needed by the unsafe code")]
+        let mut buf = [0; 32];
+        let actions = to_vec(&self.actions).unwrap();
+
+        unsafe { sys::send_proposal(Buffer::from(actions.as_slice()), BufferMut::new(buf)) }
```

Comment:
```suggestion
        unsafe { sys::send_proposal(Buffer::from(actions.as_slice()), BufferMut::new(&mut buf)) }
```

Reply:
Not needed as this is already `&mut`.

Reply:
Not needed as this is already `&mut`.

Reply:
`buf`? it's `[u8; 32]`

Reply:
So we can't do `&actions` there as a suitable `From` doesn't exist, hence the `.as_slice()` 🙂 

Meanwhile the `&mut` in `&mut buf` is unnecessary as it's already defined above as `&mut`.

Reply:
> So we can't do &actions there as a suitable From doesn't exist, hence the .as_slice() 🙂

Ah.. okay, was trying to satisfy `Buffer: From<&[u8]>`, thought it would auto-deref, anyway  `&*actions` would've done the trick.

> as it's already defined above as `&mut`.

you've declared it as mutable, but you've not gotten a mutable reference to it

Reply:
When I put it in, the compiler tells me to take it out as it's already present 🤔

Reply:
just.. follow [other](https://github.com/calimero-network/core/blob/5df5176b91708d89b6544a30370276f1d474a4c7/crates/sdk/src/env.rs#L101-L104) [references](https://github.com/calimero-network/core/blob/5df5176b91708d89b6544a30370276f1d474a4c7/crates/sdk/src/env.rs#L201-L204)

Reply:
Ahhh, I'm with you now 😄 Sorry, I hadn't made the link